File: src/main/java/com/refinedmods/refinedstorage/block/ExporterBlock.java
package com.refinedmods.refinedstorage.block;

import com.refinedmods.refinedstorage.block.shape.ShapeCache;
import com.refinedmods.refinedstorage.container.ExporterContainerMenu;
import com.refinedmods.refinedstorage.container.factory.BlockEntityMenuProvider;
import com.refinedmods.refinedstorage.render.ConstantsCable;
import com.refinedmods.refinedstorage.blockentity.ExporterBlockEntity;
import com.refinedmods.refinedstorage.util.BlockUtils;
import com.refinedmods.refinedstorage.util.CollisionUtils;
import com.refinedmods.refinedstorage.util.NetworkUtils;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;

import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;
import net.minecraftforge.network.NetworkHooks;

public class ExporterBlock extends CableBlock {
    private static final VoxelShape LINE_NORTH_1 = box(6, 6, 0, 10, 10, 2);
    private static final VoxelShape LINE_NORTH_2 = box(5, 5, 2, 11, 11, 4);
    private static final VoxelShape LINE_NORTH_3 = box(3, 3, 4, 13, 13, 6);
    private static final VoxelShape LINE_NORTH = Shapes.or(LINE_NORTH_1, LINE_NORTH_2, LINE_NORTH_3);

    private static final VoxelShape LINE_EAST_1 = box(14, 6, 6, 16, 10, 10);
    private static final VoxelShape LINE_EAST_2 = box(12, 5, 5, 14, 11, 11);
    private static final VoxelShape LINE_EAST_3 = box(10, 3, 3, 12, 13, 13);
    private static final VoxelShape LINE_EAST = Shapes.or(LINE_EAST_1, LINE_EAST_2, LINE_EAST_3);

    private static final VoxelShape LINE_SOUTH_1 = box(6, 6, 14, 10, 10, 16);
    private static final VoxelShape LINE_SOUTH_2 = box(5, 5, 12, 11, 11, 14);
    private static final VoxelShape LINE_SOUTH_3 = box(3, 3, 10, 13, 13, 12);
    private static final VoxelShape LINE_SOUTH = Shapes.or(LINE_SOUTH_1, LINE_SOUTH_2, LINE_SOUTH_3);

    private static final VoxelShape LINE_WEST_1 = box(0, 6, 6, 2, 10, 10);
    private static final VoxelShape LINE_WEST_2 = box(2, 5, 5, 4, 11, 11);
    private static final VoxelShape LINE_WEST_3 = box(4, 3, 3, 6, 13, 13);
    private static final VoxelShape LINE_WEST = Shapes.or(LINE_WEST_1, LINE_WEST_2, LINE_WEST_3);

    private static final VoxelShape LINE_UP_1 = box(6, 14, 6, 10, 16, 10);
    private static final VoxelShape LINE_UP_2 = box(5, 12, 5, 11, 14, 11);
    private static final VoxelShape LINE_UP_3 = box(3, 10, 3, 13, 12, 13);
    private static final VoxelShape LINE_UP = Shapes.or(LINE_UP_1, LINE_UP_2, LINE_UP_3);

    private static final VoxelShape LINE_DOWN_1 = box(6, 0, 6, 10, 2, 10);
    private static final VoxelShape LINE_DOWN_2 = box(5, 2, 5, 11, 4, 11);
    private static final VoxelShape LINE_DOWN_3 = box(3, 4, 3, 13, 6, 13);
    private static final VoxelShape LINE_DOWN = Shapes.or(LINE_DOWN_1, LINE_DOWN_2, LINE_DOWN_3);

    public ExporterBlock() {
        super(BlockUtils.DEFAULT_GLASS_PROPERTIES);
    }

    @Override
    public BlockDirection getDirection() {
        return BlockDirection.ANY;
    }

    @Override
    public VoxelShape getShape(BlockState state, BlockGetter world, BlockPos pos, CollisionContext ctx) {
        return ConstantsCable.addCoverVoxelShapes(ShapeCache.getOrCreate(state, s -> {
            VoxelShape shape = getCableShape(s);

            shape = Shapes.or(shape, getLineShape(s));

            return shape;
        }), world, pos);
    }


    private VoxelShape getLineShape(BlockState state) {
        Direction direction = state.getValue(getDirection().getProperty());

        if (direction == Direction.NORTH) {
            return LINE_NORTH;
        }

        if (direction == Direction.EAST) {
            return LINE_EAST;
        }

        if (direction == Direction.SOUTH) {
            return LINE_SOUTH;
        }

        if (direction == Direction.WEST) {
            return LINE_WEST;
        }

        if (direction == Direction.UP) {
            return LINE_UP;
        }

        if (direction == Direction.DOWN) {
            return LINE_DOWN;
        }

        return Shapes.empty();
    }

    @Override
    public BlockEntity newBlockEntity(BlockPos pos, BlockState state) {
        return new ExporterBlockEntity(pos, state);
    }

    @Override
    @SuppressWarnings("deprecation")
    public InteractionResult use(BlockState state, Level level, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
        if (!level.isClientSide && CollisionUtils.isInBounds(getLineShape(state), pos, hit.getLocation())) {
            return NetworkUtils.attemptModify(level, pos, player, () -> NetworkHooks.openScreen(
                (ServerPlayer) player,
                new BlockEntityMenuProvider<ExporterBlockEntity>(
                    Component.translatable("gui.refinedstorage.exporter"),
                    (blockEntity, windowId, inventory, p) -> new ExporterContainerMenu(blockEntity, player, windowId),
                    pos
                ),
                pos
            ));
        }

        return InteractionResult.SUCCESS;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/block/ExporterBlock.kt
package com.refinedmods.refinedstorage.block

import com.refinedmods.refinedstorage.RS
import com.refinedmods.refinedstorage.block.shape.ShapeCache.getOrCreate
import com.refinedmods.refinedstorage.tile.NoOpBlockEntity
import com.refinedmods.refinedstorage.util.BlockUtils
import com.thinkslynk.fabric.annotations.registry.RegisterBlock
import net.minecraft.block.BlockEntityProvider
import net.minecraft.block.BlockState
import net.minecraft.block.ShapeContext
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes
import net.minecraft.world.BlockView
import net.minecraft.world.World
import java.util.function.Function

@RegisterBlock(RS.ID, ExporterBlock.ID)
class ExporterBlock:
        CableBlock(BlockUtils.DEFAULT_GLASS_PROPERTIES, false) // TODO Check connected
//        BlockEntityProvider
{
    override val direction: BlockDirection
        get() = BlockDirection.ANY

    override fun getOutlineShape(state: BlockState, world: BlockView, pos: BlockPos, context: ShapeContext): VoxelShape {
        return getOrCreate(state, Function { s: BlockState ->
            VoxelShapes.union(getCableShape(s), getLineShape(s))
        })
    }

    private fun getLineShape(state: BlockState): VoxelShape {
        return LINE_NORTH
//        val direction: Direction = state.get(direction.property)
//        return when {
//            direction === Direction.NORTH -> LINE_NORTH
//            direction === Direction.EAST -> LINE_EAST
//            direction === Direction.SOUTH -> LINE_SOUTH
//            direction === Direction.WEST -> LINE_WEST
//            direction === Direction.UP -> LINE_UP
//            direction === Direction.DOWN -> LINE_DOWN
//            else -> VoxelShapes.empty()
//        }
    }

//    override fun createBlockEntity(world: BlockView): BlockEntity
//            = NoOpBlockEntity()
//    // TODO BlockEntities
////            = ExporterTile()

    override fun onUse(state: BlockState?, world: World?, pos: BlockPos?, player: PlayerEntity?, hand: Hand?, hit: BlockHitResult?): ActionResult {
        // TODO Port Gui
//        return if (!world.isClient && CollisionUtils.isInBounds(getLineShape(state), pos, hit.getHitVec())) {
//            NetworkUtils.attemptModify(world, pos, hit.getFace(), player) {
//                NetworkHooks.openGui(
//                        player as ServerPlayerEntity?,
//                        PositionalTileContainerProvider<ExporterTile>(
//                                TranslationTextComponent("gui.refinedstorage.exporter"),
//                                { tile: ExporterTile?, windowId: Int, inventory: PlayerInventory?, p: PlayerEntity? -> ExporterContainer(tile, player, windowId) },
//                                pos
//                        ),
//                        pos
//                )
//            }
//        } else ActionResult.SUCCESS
        return ActionResult.SUCCESS
    }

    companion object {
        const val ID = "exporter"
        private val LINE_NORTH_1: VoxelShape = createCuboidShape(6.0, 6.0, 0.0, 10.0, 10.0, 2.0)
        private val LINE_NORTH_2: VoxelShape = createCuboidShape(5.0, 5.0, 2.0, 11.0, 11.0, 4.0)
        private val LINE_NORTH_3: VoxelShape = createCuboidShape(3.0, 3.0, 4.0, 13.0, 13.0, 6.0)
        private val LINE_NORTH: VoxelShape = VoxelShapes.union(LINE_NORTH_1, LINE_NORTH_2, LINE_NORTH_3)
        private val LINE_EAST_1: VoxelShape = createCuboidShape(14.0, 6.0, 6.0, 16.0, 10.0, 10.0)
        private val LINE_EAST_2: VoxelShape = createCuboidShape(12.0, 5.0, 5.0, 14.0, 11.0, 11.0)
        private val LINE_EAST_3: VoxelShape = createCuboidShape(10.0, 3.0, 3.0, 12.0, 13.0, 13.0)
        private val LINE_EAST: VoxelShape = VoxelShapes.union(LINE_EAST_1, LINE_EAST_2, LINE_EAST_3)
        private val LINE_SOUTH_1: VoxelShape = createCuboidShape(6.0, 6.0, 14.0, 10.0, 10.0, 16.0)
        private val LINE_SOUTH_2: VoxelShape = createCuboidShape(5.0, 5.0, 12.0, 11.0, 11.0, 14.0)
        private val LINE_SOUTH_3: VoxelShape = createCuboidShape(3.0, 3.0, 10.0, 13.0, 13.0, 12.0)
        private val LINE_SOUTH: VoxelShape = VoxelShapes.union(LINE_SOUTH_1, LINE_SOUTH_2, LINE_SOUTH_3)
        private val LINE_WEST_1: VoxelShape = createCuboidShape(0.0, 6.0, 6.0, 2.0, 10.0, 10.0)
        private val LINE_WEST_2: VoxelShape = createCuboidShape(2.0, 5.0, 5.0, 4.0, 11.0, 11.0)
        private val LINE_WEST_3: VoxelShape = createCuboidShape(4.0, 3.0, 3.0, 6.0, 13.0, 13.0)
        private val LINE_WEST: VoxelShape = VoxelShapes.union(LINE_WEST_1, LINE_WEST_2, LINE_WEST_3)
        private val LINE_UP_1: VoxelShape = createCuboidShape(6.0, 14.0, 6.0, 10.0, 16.0, 10.0)
        private val LINE_UP_2: VoxelShape = createCuboidShape(5.0, 12.0, 5.0, 11.0, 14.0, 11.0)
        private val LINE_UP_3: VoxelShape = createCuboidShape(3.0, 10.0, 3.0, 13.0, 12.0, 13.0)
        private val LINE_UP: VoxelShape = VoxelShapes.union(LINE_UP_1, LINE_UP_2, LINE_UP_3)
        private val LINE_DOWN_1: VoxelShape = createCuboidShape(6.0, 0.0, 6.0, 10.0, 2.0, 10.0)
        private val LINE_DOWN_2: VoxelShape = createCuboidShape(5.0, 2.0, 5.0, 11.0, 4.0, 11.0)
        private val LINE_DOWN_3: VoxelShape = createCuboidShape(3.0, 4.0, 3.0, 13.0, 6.0, 13.0)
        private val LINE_DOWN: VoxelShape = VoxelShapes.union(LINE_DOWN_1, LINE_DOWN_2, LINE_DOWN_3)
    }
}